From e03fe6a4bec1d057f0030f2066066297363623e4 Mon Sep 17 00:00:00 2001
From: Behan Webster <behanw@converseincode.com>
Date: Wed, 12 Feb 2014 12:16:28 -0800
Subject: [PATCH] arm64: LLVMLinux: Add current_stack_pointer() for arm64

A macro to get the current stack pointer which allows for a single place in
which to do so with ASM. Before this named registers (a gcc extension) was used
to get the stack pointer. Using ASM is a more portable way of getting the stack
pointer which works with both gcc and clang.  This macro is of the same name
used in the X86 arch.

Signed-off-by: Behan Webster <behanw@converseincode.com>
---
 arch/arm64/include/asm/thread_info.h | 9 +++++++++
 1 file changed, 9 insertions(+)

--- linaro-aarch64.orig/arch/arm64/include/asm/thread_info.h
+++ linaro-aarch64/arch/arm64/include/asm/thread_info.h
@@ -69,6 +69,15 @@
 #define init_stack		(init_thread_union.stack)
 
 /*
+ * how to get the current stack pointer from C
+ */
+#define current_stack_pointer ({                    \
+	register unsigned long sp asm("sp") __used; \
+	asm("" : "=r" (sp));                        \
+	sp;                                         \
+})
+
+/*
  * how to get the thread information struct from C
  */
 static inline struct thread_info *current_thread_info(void) __attribute_const__;
